libs/patchwheel/patch.go

```diff
@@ -0,0 +1,280 @@
+// If the target wheel already exists, it returns the path to the existing wheel without redoing the patching.
+func PatchWheel(ctx context.Context, path, outputDir string) (string, error) {
+	fileInfo, err := os.Stat(path)
+	if err != nil {
+		return "", err
+	}
+	wheelMtime := fileInfo.ModTime().UTC()
+
+	filename := filepath.Base(path)
+	wheelInfo, err := ParseWheelFilename(filename)
+	if err != nil {
+		return "", err
+	}
+
+	newVersion, newFilename := CalculateNewVersion(wheelInfo, wheelMtime)
+	outpath := filepath.Join(outputDir, newFilename)
+
+	if _, err := os.Stat(outpath); err == nil {
+		// Target wheel already exists, return its path
+		return outpath, nil
+	}
+
+	tmpFilename := outpath + fmt.Sprintf(".tmp%d", os.Getpid())
+
+	needRemoval := true
+
+	defer func() {
+		if needRemoval {
+			_ = os.Remove(tmpFilename)
+		}
+	}()
+
+	r, err := zip.OpenReader(path)
+	if err != nil {
+		return "", err
+	}
+	defer r.Close()
+
+	metadataFile, recordFile, oldDistInfoPrefix := readMetadataAndRecord(r)
+	if metadataFile == nil {
+		return "", fmt.Errorf("wheel %s missing METADATA file", path)
+	}
+
+	if recordFile == nil {
+		return "", fmt.Errorf("wheel %s missing RECORD file", path)
+	}
+
+	metadataReader, err := metadataFile.Open()
+	if err != nil {
+		return "", err
+	}
+	defer metadataReader.Close()
+
+	newMetadata, err := patchMetadata(metadataReader, wheelInfo.Version, newVersion)
+	if err != nil {
+		return "", err
+	}
+
+	h := sha256.New()
+	h.Write(newMetadata)
+	metadataHash := base64.URLEncoding.WithPadding(base64.NoPadding).EncodeToString(h.Sum(nil))
+	metadataSize := len(newMetadata)
+
+	// Compute the new dist-info directory prefix.
+	newDistInfoPrefix := strings.Replace(oldDistInfoPrefix, wheelInfo.Version, newVersion, 1)
+	if newDistInfoPrefix == oldDistInfoPrefix {
+		return "", fmt.Errorf("unexpected dist-info directory format: %s (version=%s)", oldDistInfoPrefix, wheelInfo.Version)
+	}
+
+	recordReader, err := recordFile.Open()
+	if err != nil {
+		return "", err
+	}
+	defer recordReader.Close()
+
+	newRecord, err := patchRecord(recordReader, oldDistInfoPrefix, newDistInfoPrefix, metadataHash, metadataSize)
+	if err != nil {
+		return "", err
+	}
+
+	outFile, err := os.Create(tmpFilename)
+	if err != nil {
+		return "", err
+	}
+	defer outFile.Close()
+
+	metadataUpdated := 0
+	recordUpdated := 0
+
+	zipw := zip.NewWriter(outFile)
+	for _, f := range r.File {
+		// If the file is inside the old dist-info directory, update its name.
+		newName := f.Name
+		if strings.HasPrefix(f.Name, oldDistInfoPrefix) {
+			newName = newDistInfoPrefix + f.Name[len(oldDistInfoPrefix):]
+		}
+		header := &zip.FileHeader{
+			Name:   newName,
+			Method: f.Method,
+		}
+
+		header.Modified = f.ModTime()
+		header.Comment = f.Comment
+		if f.FileInfo().IsDir() && !strings.HasSuffix(header.Name, "/") {
+			header.Name += "/"
+		}
+
+		writer, err := zipw.CreateHeader(header)
```

Comment:
It seems like `writer.Close` is not called, should it be?

Reply:
According to the example here https://pkg.go.dev/archive/zip#example-Writer it is not needed. Also the docs don't mention it.